apps/enphasesolar/enphase_solar.star

```diff
@@ -0,0 +1,174 @@
+"""
+Applet: Enphase Solar
+Summary: Solar energy monitor
+Description: Display energy generated by solar panels based on Enlighten API v4.
+Author: laodaochong
+"""
+
+load("cache.star", "cache")
+load("encoding/base64.star", "base64")
+load("http.star", "http")
+load("humanize.star", "humanize")
+load("render.star", "render")
+load("schema.star", "schema")
+
+ENDPOINT = "https://api.enphaseenergy.com/api/v4/systems"
+AUTH_URL = "https://api.enphaseenergy.com/oauth/token"
+EXPIRE_MSG = "Access token has expired"
+
+# Due to limited number of API calls in free version of Enphase Application, here it only update
+# information every hour.
+TTL_SECONDS = 3600
+
+def check_response(response):
+    if response.status_code == 200:
+        return 200, response.json()
+    elif response.status_code == 401 and EXPIRE_MSG in response.json().get("details"):
+        print("Access token is expired and refresh token now. Call next time.")
+        return 401, None
+    else:
+        print("Request is failed with status code {} {}".format(response.status_code, response.json()))
+        return response.status_code, None
+
+def get_system_stats(api_key):
+    headers = {
+        "Authorization": "Bearer " + cache.get("access_token"),
```

Comment:
Cache is scoped to the app, but not to individual users. You'll need to update all cache keys so that they're unique to the user/configuration.